runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/TraceSpan.kt

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.tracing
+
+import aws.smithy.kotlin.runtime.io.Closeable
+import aws.smithy.kotlin.runtime.time.Instant
+
+/**
+ * Defines a logical lifecycle within which events may occur. Spans are typically created before some notable operation
+ * and closed after that operation is completed. Such operations may be very long and may involve many sub-operations,
+ * some of which may warrant child spans.
+ */
+public interface TraceSpan : Closeable {
+    /**
+     * The identifier for this span, which should be unique among sibling spans within the same parent. Trace span IDs
+     * may be used by probes to collate or decorate event output.
+     */
+    public val id: String
+
+    /**
+     * The parent span for this child span (if any).
+     */
+    public val parent: TraceSpan?
+
+    /**
+     * Creates a new child span with the given ID.
+     * @param id The id for the new span. IDs should be unique among sibling spans within the same parent.
+     * @return The new child span.
+     */
+    public fun child(id: String): TraceSpan
+
+    /**
+     * Records new events that have occurred within the logical context of this span.
+     * @param events The events to record.
+     */
+    public fun postEvents(events: Iterable<TraceEvent>)
+}
+
+/**
+ * Records a single new event that has occurred within the logical context of this span.
+ * @param event The event to record.
+ */
+public fun TraceSpan.postEvent(event: TraceEvent): Unit = postEvents(listOf(event))
```

Comment:
I'm guessing this is going to be our primary use case (single event). Is there any concern/thought given to the overhead of list allocation on every event?

Reply:
Giving some thought to this, yes there would be some additional overhead of list allocation. I believe that's justified by two factors:

* The receiving side of this (`TraceProbe`) also specifies a list-typed consuming method: `fun postEvents(span: TraceSpan, events: Iterable<TraceEvent>)`. If we offered a single-item `TraceSpan.postEvent` method, we'd need to support a similar single-event `TraceProbe.postEvent` method in the consumer. This would force implementors to write two methods for handling events, which seems like poor interface design. It's important to be able to support the batch case so that implementors can take advantage of scaling efficiencies in their downstream systems/logic so it also seems wrong to eliminate the list-typed overload.
* The [implementation of `listOf(element: T)`](https://github.com/JetBrains/kotlin/blob/ea836fd46a1fef07d77c96f9d7e8d7807f793453/libraries/stdlib/jvm/src/kotlin/collections/CollectionsJVM.kt#L20) on JVM delegates down to [`Collections.singletonList`](https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/util/Collections.java#L4847-L4898), which utilizes an optimized list class with very little overhead (compared to mutable list implementations). I believe the cost here is low.